src/profile-logic/profile-data.js

```diff
@@ -594,46 +595,156 @@ export function filterThreadToRange(
   });
 }
 
+// --------------- CallNodePath and CallNodeIndex manipulations ---------------
+
+// Returns a list of CallNodeIndex from CallNodePaths. This function uses a map
+// to speed up the look-up process.
 export function getCallNodeIndicesFromPaths(
   callNodePaths: CallNodePath[],
   callNodeTable: CallNodeTable
 ): Array<IndexIntoCallNodeTable | null> {
+  // This is a Map<CallNodePathHash, IndexIntoCallNodeTable>. This map speeds up
+  // the look-up process by caching every CallNodePath we handle which avoids
+  // looking up parents again and again.
+  const cache = new Map();
   return callNodePaths.map(path =>
-    getCallNodeIndexFromPath(path, callNodeTable)
+    _getCallNodeIndexFromPathWithCache(path, callNodeTable, cache)
   );
 }
 
-export function getCallNodeIndexFromPath(
+// Returns a CallNodeIndex from a CallNodePath, using and contributing to the
+// cache parameter.
+function _getCallNodeIndexFromPathWithCache(
   callNodePath: CallNodePath,
+  callNodeTable: CallNodeTable,
+  cache: Map<string, IndexIntoCallNodeTable>
+): IndexIntoCallNodeTable | null {
+  const hashFullPath = hashPath(callNodePath);
+  let result = cache.get(hashFullPath);
+  if (result !== undefined) {
+    // The cache already has the result for the full path.
+    return result;
+  }
+
+  // This map is basically the result of:
+  // `map.set(index, hashPath(callNodePath.slice(0, index)))`
+  // This is a cache to speed up contributing data to the `cache` map.
+  const sliceHashes = new Map();
+  sliceHashes.set(callNodePath.length, hashFullPath);
```

Comment:
I would prefer this map to be an array instead, indexed by `callNodePath.length - subPath.length`. This would incur less hashing and would remove the necessity to check for undefined in the per-iteration callback, where we know that this index is present in the map but have to check anyway in order to make flow happy. (At least I think that's the case, please correct me if I'm wrong.)

Reply:
Thanks, I think this is faster !

Reply:
Which makes me think that in PathSet, instead of using a Map we could use a pure object. It seems faster too... (which is sad)

Reply:
> Which makes me think that in PathSet, instead of using a Map we could use a pure object. It seems faster too... (which is sad)

(I prepared this for a separate PR, I won't include it here)